File: UVA/未分類/UVA-644 Immediate Decodability.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

class TrieNode {
public:
    TrieNode* child[2];
    int cnt;
    bool is_end;
    TrieNode() {
        child[0] = child[1] = nullptr;
        cnt = 0;
        is_end = false;
    }

    void insert(const string& s) {
        TrieNode* node = this;
        for (char ch : s) {
            int idx = ch - '0';
            if (node->child[idx] == nullptr) {
                node->child[idx] = new TrieNode();
            }
            node = node->child[idx];
        }
        node->is_end = true;
        node->cnt += 1;
    }

    bool search_prefix(const string& s) {
        TrieNode* node = this;
        for (char ch : s) {
            if (node->is_end) return true;
            int idx = ch - '0';
            node = node->child[idx];
        }
        return node->cnt > 1;
    }
};
int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int kase = 1;
    string s;
    vector<string> words;
    while (cin >> s) {
        TrieNode root = TrieNode();
        words.clear();
        root.insert(s);
        words.push_back(s);
        while (cin >> s && s != "9") {
            root.insert(s);
            words.push_back(s);
        }
        bool flag = false;
        for (string s : words) {
            if (root.search_prefix(s)) {
                flag = true;
                break;
            }
        }
        if (flag) {
            cout << "Set " << kase << " is not immediately decodable" << endl;
        } else {
            cout << "Set " << kase << " is immediately decodable" << endl;
        }
        kase++;
    }
    return 0;
}

File: UVA/未分類/UVA-644 Immediate Decodability.py
import sys
def print(val=""): sys.stdout.write(str(val) + "\n")

class TrieNode:
    def __init__(self):
        self.child = [None] * 2
        self.cnt = 0
        self.is_end = False

    def insert(self, s):
        node = self
        for ch in s:
            idx = ord(ch) - ord('0')
            if node.child[idx] is None:
                node.child[idx] = TrieNode()
            node = node.child[idx]
        node.is_end = True
        node.cnt += 1

    def search_prefix(self, s):
        node = self
        for ch in s:
            if node.is_end:
                return True
            idx = ord(ch) - ord('0')
            node = node.child[idx]
        return node.cnt > 1
kase = 1
root = TrieNode()
words = []
ans = []
for line in sys.stdin:
    line = line.strip()
    if line == "9":
        flag = False
        for word in words:
            if root.search_prefix(word):
                flag = True
                break
        if flag:
            ans.append(f"Set {kase} is not immediately decodable")
        else:
            ans.append(f"Set {kase} is immediately decodable")
        kase += 1
        # reset
        root = TrieNode()
        words = []
    else:
        root.insert(line)
        words.append(line)

print("\n".join(ans))